storage/innobase/trx/trx0rseg.cc

```diff
@@ -117,7 +118,7 @@ trx_rseg_update_wsrep_checkpoint(
 
 	if (xid_seqno != -1
 	    && !memcmp(xid_uuid, wsrep_uuid, sizeof wsrep_uuid)) {
-		ut_ad(xid_seqno > wsrep_seqno);
+		ut_ad(for_rollback || xid_seqno > wsrep_seqno);
```

Comment:
This assertion does not catch the situation where the seqno for rollback is synced too early. This could lead to situation where we have
```
commit(1) rollback(3)
```
and seqno 2 still executing. If the database crashes before seqno 2 is committed, the recovery ends up getting a wrong position. The recovery will hit this assertion when the transaction with seqno 2 gets committed during recovery.

I think it would be better to have condition before this if branch
```
if (xid_seqno != -1 && for_rollback && xid_seqno < wsrep_seqno)
{
  return;
}
```
to avoid syncing the seqno if out of date. The assertion could be then left as it was before. And maybe this check/early return should also be in optimized builds.